### DISCUSSION TITLE
Why this error "Cypress detected that you returned a promise from a command while also invoking one or more cy commands in that promise" happen only when running the test remotely using Jenkins?

### DISCUSSION THREAD

**Discussion by TerryMeirC** [cypress-io/cypress#24793] [Questions and Help]
Why this error "Cypress detected that you returned a promise from a command while also invoking one or more cy commands in that promise" happen only when running the test remotely using Jenkins?

My code uses a recursive function as an implementation of the while loop. the recursive function returns a promise to ensure the returning value (function search type, after finding what was searched returning the value).

When I ran the test locally the test passed with no errors, but remotely using Jenkins I'm getting the error "Cypress detected that you returned a promise from a command while also invoking one or more cy commands in that promise"

**Comment by portyu9** [comment]
Hi @TerryMeirC 

Jenkins is most likely **exposing the code path/timing that contains the problem**, rather than being the cause of the error.

The important part is that the recursive function returns a native `Promise` **and also runs Cypress commands inside that Promise**. Cypress explicitly does not support that pattern because Cypress commands are already queued asynchronously.

Instead of:

```js
return new Promise((resolve) => {
  cy.get(...).then(() => {
    // ...
    resolve(value)
  })
})
```

return the Cypress chain itself:

```js
function searchUntilFound() {
  return cy.get('...').then(($el) => {
    if (isWhatWeWant($el)) {
      return $el
    }

    return searchUntilFound()
  })
}
```

The same pattern works for polling requests:

```js
function poll() {
  return cy.request('/some-endpoint').then((response) => {
    if (response.body.ready) {
      return response
    }

    return poll()
  })
}
```

A native Promise is fine for genuinely external asynchronous work, but don't enqueue `cy.*` commands from inside that Promise.

Since the recursive function itself wasn't included in the post, I can't point to its exact offending line, but the Promise/Cypress-command mixing described in the question is exactly what that Cypress error reports.

Please mark this answer as accepted if it helped, thank you.
